File: year2020/day02/python/solution.py
def main():
    print(part_1())
    print(part_2())


def part_1():
    sum = 0
    for line in input:
        borders, char, password = line
        min_count, max_count = map(int, borders.split('-'))

        if min_count <= password.count(char.rstrip(':')) <= max_count:
            sum += 1
            
    return sum


def part_2():
    sum = 0
    for line in input:
        borders, char, password = line
        first_pos, second_pos = map(int, borders.split('-'))

        if (password[first_pos - 1] == char.rstrip(':')) != (password[second_pos - 1] == char.rstrip(':')):
            sum += 1

    return sum


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = [int(row) for row in file]


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()

File: year2020/day02/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

public class Solution {
  private static List<String[]> PASSWORDS;

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    int count = 0;
    for (String[] pw : PASSWORDS) {
      int chars = countOccurrences(pw[2], pw[1].charAt(0));
      if (Integer.parseInt(pw[0].split("-")[0]) <= chars && chars <= Integer.parseInt(pw[0].split("-")[1])) {
        count++;
      }
    }
    return count;
  }

  public static int part2() {
    int count = 0;
    for (String[] pw : PASSWORDS) {
      if ((pw[2].charAt(Integer.parseInt(pw[0].split("-")[0]) - 1) == pw[1].charAt(0)) != (pw[2].charAt(Integer.parseInt(pw[0].split("-")[1]) - 1) == pw[1].charAt(0))) {
        count++;
      }
    }
    return count;
  }

  public static void setup(String input) {
    PASSWORDS = Arrays.stream(input.split("\n")).map(password -> password.split(" ")).collect(Collectors.toList());
  }

  private static int countOccurrences(String str, char target) {
    int count = 0;
    for (char c : str.toCharArray()) {
      if (c == target) {
        count++;
      }
    }
    return count;
  }
}